File: wpilibj/src/main/java/edu/wpi/first/wpilibj/simulation/RoboRioSim.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj.simulation;

import edu.wpi.first.hal.simulation.NotifyCallback;
import edu.wpi.first.hal.simulation.RoboRioDataJNI;
import edu.wpi.first.wpilibj.RobotController;

/** A utility class to control a simulated RoboRIO. */
public final class RoboRioSim {
  private RoboRioSim() {
    // Utility class
  }

  /**
   * Register a callback to be run whenever the Vin voltage changes.
   *
   * @param callback the callback
   * @param initialNotify whether to call the callback with the initial state
   * @return the {@link CallbackStore} object associated with this callback.
   */
  public static CallbackStore registerVInVoltageCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = RoboRioDataJNI.registerVInVoltageCallback(callback, initialNotify);
    return new CallbackStore(uid, RoboRioDataJNI::cancelVInVoltageCallback);
  }

  /**
   * Measure the Vin voltage.
   *
   * @return the Vin voltage
   */
  public static double getVInVoltage() {
    return RoboRioDataJNI.getVInVoltage();
  }

  /**
   * Define the Vin voltage.
   *
   * @param vInVoltage the new voltage
   */
  public static void setVInVoltage(double vInVoltage) {
    RoboRioDataJNI.setVInVoltage(vInVoltage);
  }

  /**
   * Register a callback to be run whenever the Vin current changes.
   *
   * @param callback the callback
   * @param initialNotify whether the callback should be called with the initial value
   * @return the {@link CallbackStore} object associated with this callback.
   */
  public static CallbackStore registerVInCurrentCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = RoboRioDataJNI.registerVInCurrentCallback(callback, initialNotify);
    return new CallbackStore(uid, RoboRioDataJNI::cancelVInCurrentCallback);
  }

  /**
   * Measure the Vin current.
   *
   * @return the Vin current
   */
  public static double getVInCurrent() {
    return RoboRioDataJNI.getVInCurrent();
  }

  /**
   * Define the Vin current.
   *
   * @param vInCurrent the new current
   */
  public static void setVInCurrent(double vInCurrent) {
    RoboRioDataJNI.setVInCurrent(vInCurrent);
  }

  /**
   * Register a callback to be run whenever the 3.3V rail voltage changes.
   *
   * @param callback the callback
   * @param initialNotify whether the callback should be called with the initial value
   * @return the {@link CallbackStore} object associated with this callback.
   */
  public static CallbackStore registerUserVoltage3V3Callback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = RoboRioDataJNI.registerUserVoltage3V3Callback(callback, initialNotify);
    return new CallbackStore(uid, RoboRioDataJNI::cancelUserVoltage3V3Callback);
  }

  /**
   * Measure the 3.3V rail voltage.
   *
   * @return the 3.3V rail voltage
   */
  public static double getUserVoltage3V3() {
    return RoboRioDataJNI.getUserVoltage3V3();
  }

  /**
   * Define the 3.3V rail voltage.
   *
   * @param userVoltage3V3 the new voltage
   */
  public static void setUserVoltage3V3(double userVoltage3V3) {
    RoboRioDataJNI.setUserVoltage3V3(userVoltage3V3);
  }

  /**
   * Register a callback to be run whenever the 3.3V rail current changes.
   *
   * @param callback the callback
   * @param initialNotify whether the callback should be called with the initial value
   * @return the {@link CallbackStore} object associated with this callback.
   */
  public static CallbackStore registerUserCurrent3V3Callback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = RoboRioDataJNI.registerUserCurrent3V3Callback(callback, initialNotify);
    return new CallbackStore(uid, RoboRioDataJNI::cancelUserCurrent3V3Callback);
  }

  /**
   * Measure the 3.3V rail current.
   *
   * @return the 3.3V rail current
   */
  public static double getUserCurrent3V3() {
    return RoboRioDataJNI.getUserCurrent3V3();
  }

  /**
   * Define the 3.3V rail current.
   *
   * @param userCurrent3V3 the new current
   */
  public static void setUserCurrent3V3(double userCurrent3V3) {
    RoboRioDataJNI.setUserCurrent3V3(userCurrent3V3);
  }

  /**
   * Register a callback to be run whenever the 3.3V rail active state changes.
   *
   * @param callback the callback
   * @param initialNotify whether the callback should be called with the initial state
   * @return the {@link CallbackStore} object associated with this callback.
   */
  public static CallbackStore registerUserActive3V3Callback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = RoboRioDataJNI.registerUserActive3V3Callback(callback, initialNotify);
    return new CallbackStore(uid, RoboRioDataJNI::cancelUserActive3V3Callback);
  }

  /**
   * Get the 3.3V rail active state.
   *
   * @return true if the 3.3V rail is active
   */
  public static boolean getUserActive3V3() {
    return RoboRioDataJNI.getUserActive3V3();
  }

  /**
   * Set the 3.3V rail active state.
   *
   * @param userActive3V3 true to make rail active
   */
  public static void setUserActive3V3(boolean userActive3V3) {
    RoboRioDataJNI.setUserActive3V3(userActive3V3);
  }

  /**
   * Register a callback to be run whenever the 3.3V rail number of faults changes.
   *
   * @param callback the callback
   * @param initialNotify whether the callback should be called with the initial value
   * @return the {@link CallbackStore} object associated with this callback.
   */
  public static CallbackStore registerUserFaults3V3Callback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = RoboRioDataJNI.registerUserFaults3V3Callback(callback, initialNotify);
    return new CallbackStore(uid, RoboRioDataJNI::cancelUserFaults3V3Callback);
  }

  /**
   * Get the 3.3V rail number of faults.
   *
   * @return number of faults
   */
  public static int getUserFaults3V3() {
    return RoboRioDataJNI.getUserFaults3V3();
  }

  /**
   * Set the 3.3V rail number of faults.
   *
   * @param userFaults3V3 number of faults
   */
  public static void setUserFaults3V3(int userFaults3V3) {
    RoboRioDataJNI.setUserFaults3V3(userFaults3V3);
  }

  /**
   * Register a callback to be run whenever the Brownout voltage changes.
   *
   * @param callback the callback
   * @param initialNotify whether to call the callback with the initial state
   * @return the {@link CallbackStore} object associated with this callback.
   */
  public static CallbackStore registerBrownoutVoltageCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = RoboRioDataJNI.registerBrownoutVoltageCallback(callback, initialNotify);
    return new CallbackStore(uid, RoboRioDataJNI::cancelBrownoutVoltageCallback);
  }

  /**
   * Measure the Brownout voltage.
   *
   * @return the Brownout voltage
   */
  public static double getBrownoutVoltage() {
    return RoboRioDataJNI.getBrownoutVoltage();
  }

  /**
   * Define the Brownout voltage.
   *
   * @param vInVoltage the new voltage
   */
  public static void setBrownoutVoltage(double vInVoltage) {
    RoboRioDataJNI.setBrownoutVoltage(vInVoltage);
  }

  /**
   * Register a callback to be run whenever the cpu temp changes.
   *
   * @param callback the callback
   * @param initialNotify whether to call the callback with the initial state
   * @return the {@link CallbackStore} object associated with this callback.
   */
  public static CallbackStore registerCPUTempCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = RoboRioDataJNI.registerCPUTempCallback(callback, initialNotify);
    return new CallbackStore(uid, RoboRioDataJNI::cancelCPUTempCallback);
  }

  /**
   * Get the cpu temp.
   *
   * @return the cpu temp.
   */
  public static double getCPUTemp() {
    return RoboRioDataJNI.getCPUTemp();
  }

  /**
   * Set the cpu temp.
   *
   * @param cpuTemp the new cpu temp.
   */
  public static void setCPUTemp(double cpuTemp) {
    RoboRioDataJNI.setCPUTemp(cpuTemp);
  }

  /**
   * Register a callback to be run whenever the team number changes.
   *
   * @param callback the callback
   * @param initialNotify whether to call the callback with the initial state
   * @return the {@link CallbackStore} object associated with this callback.
   */
  public static CallbackStore registerTeamNumberCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid = RoboRioDataJNI.registerTeamNumberCallback(callback, initialNotify);
    return new CallbackStore(uid, RoboRioDataJNI::cancelTeamNumberCallback);
  }

  /**
   * Get the team number.
   *
   * @return the team number.
   */
  public static int getTeamNumber() {
    return RoboRioDataJNI.getTeamNumber();
  }

  /**
   * Set the team number.
   *
   * @param teamNumber the new team number.
   */
  public static void setTeamNumber(int teamNumber) {
    RoboRioDataJNI.setTeamNumber(teamNumber);
  }

  /**
   * Get the serial number.
   *
   * @return The serial number.
   */
  public static String getSerialNumber() {
    return RoboRioDataJNI.getSerialNumber();
  }

  /**
   * Set the serial number.
   *
   * @param serialNumber The serial number.
   */
  public static void setSerialNumber(String serialNumber) {
    RoboRioDataJNI.setSerialNumber(serialNumber);
  }

  /**
   * Get the comments string.
   *
   * @return The comments string.
   */
  public static String getComments() {
    return RoboRioDataJNI.getComments();
  }

  /**
   * Set the comments string.
   *
   * @param comments The comments string.
   */
  public static void setComments(String comments) {
    RoboRioDataJNI.setComments(comments);
  }

  /** Reset all simulation data. */
  public static void resetData() {
    RoboRioDataJNI.resetData();
  }
}

File: wpilibc/src/main/native/cpp/simulation/RoboRioSim.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/simulation/RoboRioSim.hpp"

#include <memory>
#include <string>

#include "wpi/hal/simulation/RoboRioData.h"
#include "wpi/util/string.hpp"

using namespace wpi;
using namespace wpi::sim;

std::unique_ptr<CallbackStore> RoboRioSim::RegisterVInVoltageCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      -1, callback, &HALSIM_CancelRoboRioVInVoltageCallback);
  store->SetUid(HALSIM_RegisterRoboRioVInVoltageCallback(
      &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

wpi::units::volts<> RoboRioSim::GetVInVoltage() {
  return wpi::units::volts<>{HALSIM_GetRoboRioVInVoltage()};
}

void RoboRioSim::SetVInVoltage(wpi::units::volts<> vInVoltage) {
  HALSIM_SetRoboRioVInVoltage(vInVoltage.value());
}

std::unique_ptr<CallbackStore> RoboRioSim::RegisterUserVoltage3V3Callback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      -1, callback, &HALSIM_CancelRoboRioUserVoltage3V3Callback);
  store->SetUid(HALSIM_RegisterRoboRioUserVoltage3V3Callback(
      &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

wpi::units::volts<> RoboRioSim::GetUserVoltage3V3() {
  return wpi::units::volts<>{HALSIM_GetRoboRioUserVoltage3V3()};
}

void RoboRioSim::SetUserVoltage3V3(wpi::units::volts<> userVoltage3V3) {
  HALSIM_SetRoboRioUserVoltage3V3(userVoltage3V3.value());
}

std::unique_ptr<CallbackStore> RoboRioSim::RegisterUserCurrent3V3Callback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      -1, callback, &HALSIM_CancelRoboRioUserCurrent3V3Callback);
  store->SetUid(HALSIM_RegisterRoboRioUserCurrent3V3Callback(
      &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

wpi::units::amperes<> RoboRioSim::GetUserCurrent3V3() {
  return wpi::units::amperes<>{HALSIM_GetRoboRioUserCurrent3V3()};
}

void RoboRioSim::SetUserCurrent3V3(wpi::units::amperes<> userCurrent3V3) {
  HALSIM_SetRoboRioUserCurrent3V3(userCurrent3V3.value());
}

std::unique_ptr<CallbackStore> RoboRioSim::RegisterUserActive3V3Callback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      -1, callback, &HALSIM_CancelRoboRioUserActive3V3Callback);
  store->SetUid(HALSIM_RegisterRoboRioUserActive3V3Callback(
      &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

bool RoboRioSim::GetUserActive3V3() {
  return HALSIM_GetRoboRioUserActive3V3();
}

void RoboRioSim::SetUserActive3V3(bool userActive3V3) {
  HALSIM_SetRoboRioUserActive3V3(userActive3V3);
}

std::unique_ptr<CallbackStore> RoboRioSim::RegisterUserFaults3V3Callback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      -1, callback, &HALSIM_CancelRoboRioUserFaults3V3Callback);
  store->SetUid(HALSIM_RegisterRoboRioUserFaults3V3Callback(
      &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

int RoboRioSim::GetUserFaults3V3() {
  return HALSIM_GetRoboRioUserFaults3V3();
}

void RoboRioSim::SetUserFaults3V3(int userFaults3V3) {
  HALSIM_SetRoboRioUserFaults3V3(userFaults3V3);
}

std::unique_ptr<CallbackStore> RoboRioSim::RegisterBrownoutVoltageCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      -1, callback, &HALSIM_CancelRoboRioBrownoutVoltageCallback);
  store->SetUid(HALSIM_RegisterRoboRioBrownoutVoltageCallback(
      &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

wpi::units::volts<> RoboRioSim::GetBrownoutVoltage() {
  return wpi::units::volts<>{HALSIM_GetRoboRioBrownoutVoltage()};
}

void RoboRioSim::SetBrownoutVoltage(wpi::units::volts<> vInVoltage) {
  HALSIM_SetRoboRioBrownoutVoltage(vInVoltage.value());
}

std::unique_ptr<CallbackStore>
RoboRioSim::RegisterBrownoutRecoveryVoltageCallback(NotifyCallback callback,
                                                    bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      -1, callback, &HALSIM_CancelRoboRioBrownoutRecoveryVoltageCallback);
  store->SetUid(HALSIM_RegisterRoboRioBrownoutRecoveryVoltageCallback(
      &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

wpi::units::volts<> RoboRioSim::GetBrownoutRecoveryVoltage() {
  return wpi::units::volts<>{HALSIM_GetRoboRioBrownoutRecoveryVoltage()};
}

void RoboRioSim::SetBrownoutRecoveryVoltage(
    wpi::units::volts<> brownoutRecoveryVoltage) {
  HALSIM_SetRoboRioBrownoutRecoveryVoltage(brownoutRecoveryVoltage.value());
}

std::unique_ptr<CallbackStore> RoboRioSim::RegisterCPUTempCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      -1, callback, &HALSIM_CancelRoboRioCPUTempCallback);
  store->SetUid(HALSIM_RegisterRoboRioCPUTempCallback(
      &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

wpi::units::celsius<> RoboRioSim::GetCPUTemp() {
  return wpi::units::celsius<>{HALSIM_GetRoboRioCPUTemp()};
}

void RoboRioSim::SetCPUTemp(wpi::units::celsius<> cpuTemp) {
  HALSIM_SetRoboRioCPUTemp(cpuTemp.value());
}

std::unique_ptr<CallbackStore> RoboRioSim::RegisterTeamNumberCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      -1, callback, &HALSIM_CancelRoboRioTeamNumberCallback);
  store->SetUid(HALSIM_RegisterRoboRioTeamNumberCallback(
      &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

int32_t RoboRioSim::GetTeamNumber() {
  return HALSIM_GetRoboRioTeamNumber();
}

void RoboRioSim::SetTeamNumber(int32_t teamNumber) {
  HALSIM_SetRoboRioTeamNumber(teamNumber);
}

std::string RoboRioSim::GetSerialNumber() {
  WPI_String serialNum;
  HALSIM_GetRoboRioSerialNumber(&serialNum);
  std::string serial{wpi::util::to_string_view(&serialNum)};
  WPI_FreeString(&serialNum);
  return serial;
}

void RoboRioSim::SetSerialNumber(std::string_view serialNumber) {
  auto str = wpi::util::make_string(serialNumber);
  HALSIM_SetRoboRioSerialNumber(&str);
}

std::string RoboRioSim::GetComments() {
  WPI_String comments;
  HALSIM_GetRoboRioComments(&comments);
  std::string serial{wpi::util::to_string_view(&comments)};
  WPI_FreeString(&comments);
  return serial;
}

void RoboRioSim::SetComments(std::string_view comments) {
  auto str = wpi::util::make_string(comments);
  HALSIM_SetRoboRioComments(&str);
}

void RoboRioSim::ResetData() {
  HALSIM_ResetRoboRioData();
}
